src/Adapter/Product/Image/Repository/ProductImageMultiShopRepository.php

```diff
@@ -230,6 +275,27 @@ public function deleteFromShops(ImageId $imageId, array $shopIds): void
         );
     }
 
+    /**
+     * @param ImageId $imageId
+     * @param ShopConstraint $shopConstraint
+     *
+     * @return void
+     */
+    public function deleteByShopConstraint(ImageId $imageId, ShopConstraint $shopConstraint): void
+    {
+        $shopIds = $this->getAssociatedShopIdsByShopConstraint($imageId, $shopConstraint);
+        if (empty($shopIds)) {
+            return;
+        }
+
+        $this->deleteObjectModelFromShops(
+        // We fetch the image from first shop, the values don't matter anyway we just need an Image instance
+            $this->get($imageId, reset($shopIds)),
+            $shopIds,
+            CannotDeleteProductImageException::class
+        );
+    }
```

Comment:
Hello @jolelievre, I'm not sure if this is the responsibility of the repository, could we move this into a specific command? In another PR of course, because I see that this problem was not introduced by this PR.

Reply:
The recent repositories we added don't really follow the Repository pattern, but they are all our centralized service to handle DB access whether it be via ObjectModel manipulation or sometimes via DBAL queries (sometimes a mix of both as well). So we won't change this now.

Besides, it shouldn't be in a command handler, the very purpose of those repository services is to take most of the code OUT of the command handlers, so that all the business logic and intelligence is not completely stuck inside these handlers, thus making it unusable outside of the handler.

By externalizing all this intelligence we make it easier to be reused in different handlers, we limit the code duplication and we factorize it so that it's easier to maintain and fix.